### ISSUE TITLE
Dynamic spritesheet

### ISSUE THREAD

**Issue by gabry2105** [pixijs/pixi.js#2284]
Hi, 

I'm developing an application that will show a lot of images (~10000 images), I developed a first version as follow: for each image an Image objecet is created and loaded through the src property. When completed I create a texture from that image to use in a sprite to put in a container. It works but with chrome is too slow. I read about the ParticleContainer that can be used only if each texture has the same base texture, so I change my code: 
 * At begin a base texture is created from an empty canvas.
```javascript
var canvas = document.createElement("canvas");

//set dimension according to configuration.
// ....

baseTexture = new PIXI.BaseTexture(canvas);
```
 * When application run each image is loaded as follow: 

```javascript
var image = new Image();
image.onload = onLoadFunction;
image.src = imageUrl;
```
* When the load complete the image is drawn in the canvas at coordinate x, y:

```javascript
var ctx = canvas.getContext("2d");
ctx.drawImage(image, x, y, w, y);
```
* The base texture is updated and new one is created:

```javascript
baseTexture.update();
var texture = new PIXI.Texture(baseTexture,  new PIXI.Rectangle(x, y, w, y));
```

Each time a small set of texture are loaded (i set this number to 5) the container is rendered. When the container is rendered the previous sprites disappear and I see only the last 5. At the last render operation the container becomes white. 
Where did I go wrong?

**Comment by englercj** [comment]
Do you have an example I can debug? What you posted seems fine (depending on the size of the canvas).

**Comment by gabry2105** [comment]
The code is complex, I extract the class used to create and get the textures:

```javascript
window.DynamicSpriteSheet = function() {
        /** Size of the images. */
        var size = {
            height: 32,
            width: 32
        };
        /**
        * Number of textures conntained. Start from 1 because there is the
        * empty texture.
        */
        var length = 1;
        /**
        * Number of texture that can be contained.
        */
        var capacity = 5;
        /**
         * Textures contained indexed by the url of the image.
         */
        var textures = {};
        /**
         * Canvas used to render the images added.
         */
        var canvas = document.createElement("canvas");
        /**
         * Number of 'columns' of the canvas. Images are put in the canvas as 
         * a grid. 
         */
        var columnsOnCanvas = 5;
        /** Base texture. */
        var baseTexture = PIXI.BaseTexture.fromCanvas(this.canvas);
        /** Empty texture. */
        var emptyTexture = null;

        // Start Initialization
        canvas.width = size.width;
        canvas.height = size.height;
        emptyTexture = new PIXI.Texture(
            baseTexture,
            new PIXI.Rectangle(0, 0, size.width, size.height));
        // End Intialization

        /**
        * Add an image to the sprite sheet and create the relative
        * texture (stored in textures object).
        * @param {Image} img - Standard javascript image object to add.
        */
        this.addImage = function(img) {
            var imageUrl = img.src,
                x = (length % columnsOnCanvas) * size.width,
                y = canvas.height - size.height,
                ctx = canvas.getContext("2d");

            ++length;
            //If the canvas is full it is resized.
            if (capacity === length) {
                canvas.height = canvas.height + size.height;
                capacity = capacity + columnsOnCanvas;
            }

            ctx.drawImage(img, x, y, size.width, size.height);
            baseTexture.update();
            textures[imageUrl] = new PIXI.Texture(
                baseTexture,
                new PIXI.Rectangle(x, y, size.width, size.height));
        };
        /**
         * Gets the texture according to the url passed as parameter. If the 
         * texture is not in the sprite sheet, this method return null.
         * @param {String} imageUrl  - Url of the image of the texture. It is 
         *                             used as key of the collection.
         * @return {PIXI.Texture|null} Return the texture indexed by the 
         *                             url passed parameter. If the textures 
         *                             object does not contains the url passed 
         *                             return null. 
         */
        this.getTexture = function(imageUrl) {
            return  this.textures[imageUrl] || null;
        }
        /**
         * Gets the empty texture.
         * @returns {PIXI.Texture} Empty texture. 
         */
        this.getEmptyTexture = function() {
            return emptyTexture;
        }
    }

```
The application make an instance of this class and every time an image is loaded (onLoadFunction called) it calls the addImage of the sprite sheet, then create the sprite getting the texture throught the getTexture method. At the end the sprite created is added to the container.
I thing the error is in that class because is the only thing i changed from the previous working version. In the previous version there was an other class: TextureRepo with the same three methods: addImage, getEmptyTexture and getTexture implemented as described in the first post.

**Comment by englercj** [comment]
I'd really like to see something running so I can debug it, mostly this seems OK except for a few things that I assume are copy paste errors.

Like:

```js
        /**
         * Canvas used to render the images added.
         */
        var canvas = document.createElement("canvas");

// ...

        /** Base texture. */
        var baseTexture = PIXI.BaseTexture.fromCanvas(this.canvas);
```

Also, I wouldn't get the 2d context everytime you add an image. I would just grab it once when you create the canvas.

**Comment by gabry2105** [comment]
**I found the problem: the resize of canvas**. 
I tried to remove the resize of canvas and the code works but when it is full it "override" the last image showing the wrong image in the result. So I calculate the final width and height having the number of images to render and the size of each one (32x32). This solution does not works because the max tile size of google chrome is about 2048x2048
At this point I change every thing and final working solution is the following:

1. Create only one canvas with the size set to 2048x2048 (or MAX_TEXTURE_SIZE constant)
2. When a new image is loaded put it in the canvas until it is full.
3. When canvas is full, create a new canvas of same size and a new ParticleContainer. 
4. Draw the image in the new canvas and draw the relative sprite in the new ParticleContainer. 

This solution require a lot of work to handle the diferrent containers but it is very performant.

Thank's a lot for the support.
